README.md
# [Django Dynamic Services](https://github.com/app-generator/django-dynamic-services) 

Open-source **Django** project that showcases the **[API Generator](https://appseed.us/developer-tools/django-api-generator/)** and other **[Dynamic Services](https://appseed.us/developer-tools/django-dynamic-datatables/)** - actively supported by [AppSeed](https://appseed.us/).

> 👉 [Django - Build Services without Coding](https://www.youtube.com/watch?v=EtMCK5AmdQI) - `video presentation` (learn how to use this starter)

### Features

- ✅ `Up-to-date Dependencies`
- ✅ Theme: [Django AdminLTE](https://github.com/app-generator/django-admin-adminlte) `v3.2.0`
- ✅ [API Generator](https://github.com/app-generator/django-api-generator) - `Free Python Library` 
- ✅ [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) - `Free Python Library`
- 🚀 `Deployment`: Docker, `CI/CD` flow via `Render`

<br />

![Django Dynamic Services - free developer tools provided by AppSeed.](https://user-images.githubusercontent.com/51070104/219007148-34cefb52-92cc-4ef3-bbf9-f8ee6ae2389b.png)

<br />

## Start the app in Docker

> 👉 **Step 1** - Download the code from the GH repository (using `GIT`)

```bash
$ git clone https://github.com/app-generator/django-dynamic-services.git
$ cd django-dynamic-services
```

> 👉 **Step 2** - Start the APP in `Docker`

```bash
$ docker-compose up --build
```

Visit `http://localhost:5085` in your browser. The app should be up & running.


> 👉 **Step 3** - Create Superuser in `Docker`

```bash
$ # List containes & get the ID
$ docker container ls  
$ # Create the superuser
$ docker exec <CONTAINER_ID> python manage.py createsuperuser 
```

<br />

## Manual Build 

> 👉 Download the code  

```bash
$ git clone https://github.com/app-generator/django-dynamic-services.git
$ cd django-dynamic-services
```

<br />

> 👉 Install modules via `VENV`  

```bash
$ virtualenv env
$ source env/bin/activate
$ pip install -r requirements.txt
```

<br />

> 👉 Set Up Database

```bash
$ python manage.py makemigrations
$ python manage.py migrate
```

<br />

> 👉 Generate API

```bash
$ python manage.py generate-api     # requires confirmation
// 
$ python manage.py generate-api -f  # no input (API folder is overwritten)
```

<br />

> 👉 Create the Superuser

```bash
$ python manage.py createsuperuser
```

<br />

> 👉 Start the app

```bash
$ python manage.py runserver
```

At this point, the app runs at `http://127.0.0.1:8000/` and the generated API can be found at: 

- http://localhost:8000/api/product/ - For `products`
- http://localhost:8000/api/sales/ - For `sales` 

The default API nodes can be tested via this [POSTMAN](./media/test.postman_collection) Collection.  

<br />

## How to use the API

- Start the app
- Make sure the endpoints are up & running 
- Authenticate via API and het the access token
  - `http://localhost:8000/login/jwt/` usind existing credentials 
  - Save the token in the requests `HEADER`
- `GET Requests` are public (no token required)
  - GET ALL products: `http://localhost:8000/api/product/`
  - GET product by ID: `http://localhost:8000/api/product/1/`
  - GET ALL Sales: `http://localhost:8000/api/sales/`
- `Create`, `Delete`, `Update` requires a token in the header

For API sample requests, open and edit the [POSTMAN](./media/test.postman_collection) Collection sample.

<br />

## How Update the API

- Define or update your models
- Migrate the database
- Update the configuration `API_GENERATOR` section
- Regenerate the API
  - `python manage.py generate-api`

At this point, you should be able to use the API. For more information regarding the library used to generate the code, access: 

> 👉 [API Generator for Django](https://github.com/app-generator/django-api-generator) - Open-Source Library

<br />

## Codebase structure

The project is coded using a simple and intuitive structure presented below:

```bash
< PROJECT ROOT >
   |
   |-- core/                            
   |    |-- settings.py                  # Project Configuration  
   |    |-- urls.py                      # Project Routing
   |
   |-- home/
   |    |-- views.py                     # APP Views 
   |    |-- urls.py                      # APP Routing
   |    |-- models.py                    # APP Models 
   |    |-- tests.py                     # Tests  
   |    |-- templates/                   # Theme Customisation 
   |         |-- includes                # 
   |              |-- footer.py          # Custom Footer      
   |     
   |-- requirements.txt                  # Project Dependencies
   |
   |-- env.sample                        # ENV Configuration (default values)
   |-- manage.py                         # Start the app - Django default start script
   |
   |-- ************************************************************************
```

<br /> 

---
[Django Dynamic Services](https://github.com/app-generator/django-dynamic-services) - Open-Source **Django** starter provided by **[AppSeed](https://appseed.us/)**


CHANGELOG.md
# Change Log

## [1.0.6] 2023-03-08
### Changes

- Bump [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) `1.0.21`

## [1.0.5] 2023-03-06
### Changes

- Bump [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) `1.0.16`

## [1.0.4] 2023-03-04
### Changes

- Bump [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) `1.0.12`

## [1.0.3] 2023-02-15
### Changes

- Update CI/CD flow
  - Migrate latest DB changes 
  - Generate the API

## [1.0.2] 2023-02-15
### Changes

- Bump Versions
 - [API Generator](https://github.com/app-generator/django-api-generator) `1.0.10`
 - [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) `1.0.8`


## [1.0.1] 2023-02-15
### Changes

- Integrate [Dynamic DataTables](https://github.com/app-generator/django-dynamic-datatb) 
  - `Free Python Library`

## [1.0.0] 2023-02-14
### Changes

- Initial Codebase
- Stable Version. Tested with two linked Models
  - `Product`
  - `Sales`: linked to Product (one-to-one)


core/settings.py
"""
Django settings for core project.

Generated by 'django-admin startproject' using Django 4.1.2.

For more information on this file, see
https://docs.djangoproject.com/en/4.1/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/4.1/ref/settings/
"""

import os, random, string, inspect
from pathlib import Path
from dotenv import load_dotenv
import django_dyn_dt

load_dotenv()  # take environment variables from .env.

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/4.1/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get('SECRET_KEY')
if not SECRET_KEY:
    SECRET_KEY = ''.join(random.choice( string.ascii_lowercase  ) for i in range( 32 ))

# Render Deployment Code
DEBUG = 'RENDER' not in os.environ

# Docker HOST
ALLOWED_HOSTS = ['localhost', '127.0.0.1']

# Add here your deployment HOSTS
CSRF_TRUSTED_ORIGINS = ['http://localhost:8000', 'http://localhost:5085', 'http://127.0.0.1:8000', 'http://127.0.0.1:5085']

#Render Context
RENDER_EXTERNAL_HOSTNAME = os.environ.get('RENDER_EXTERNAL_HOSTNAME')
if RENDER_EXTERNAL_HOSTNAME:    
    ALLOWED_HOSTS.append(RENDER_EXTERNAL_HOSTNAME)

# Application definition

INSTALLED_APPS = [
    'admin_adminlte.apps.AdminAdminlteConfig',
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",

    'django_api_gen',            # Django API GENERATOR  # <-- NEW
    'rest_framework',            # Include DRF           # <-- NEW 
    'rest_framework.authtoken',  # Include DRF Auth      # <-- NEW  

    'django_dyn_dt',             # <-- NEW App

    "home",
]

MIDDLEWARE = [
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "core.urls"

HOME_TEMPLATES   = os.path.join(BASE_DIR, 'home', 'templates')
DATATB_TEMPLATES = os.path.join(BASE_DIR, "django_dyn_dt/templates")

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [HOME_TEMPLATES, DATATB_TEMPLATES],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "core.wsgi.application"


# Database
# https://docs.djangoproject.com/en/4.1/ref/settings/#databases

DB_ENGINE   = os.getenv('DB_ENGINE'   , None)
DB_USERNAME = os.getenv('DB_USERNAME' , None)
DB_PASS     = os.getenv('DB_PASS'     , None)
DB_HOST     = os.getenv('DB_HOST'     , None)
DB_PORT     = os.getenv('DB_PORT'     , None)
DB_NAME     = os.getenv('DB_NAME'     , None)

if DB_ENGINE and DB_NAME and DB_USERNAME:
    DATABASES = { 
      'default': {
        'ENGINE'  : 'django.db.backends.' + DB_ENGINE, 
        'NAME'    : DB_NAME,
        'USER'    : DB_USERNAME,
        'PASSWORD': DB_PASS,
        'HOST'    : DB_HOST,
        'PORT'    : DB_PORT,
        }, 
    }
else:
    DATABASES = {
        'default': {
            'ENGINE': 'django.db.backends.sqlite3',
            'NAME': 'db.sqlite3',
        }
    }

# Password validation
# https://docs.djangoproject.com/en/4.1/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]


# Internationalization
# https://docs.djangoproject.com/en/4.1/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "UTC"

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/4.1/howto/static-files/

STATIC_URL = '/static/'
STATIC_ROOT = os.path.join(BASE_DIR, 'staticfiles')

#if not DEBUG:
#    STATICFILES_STORAGE = 'whitenoise.storage.CompressedManifestStaticFilesStorage'

# Default primary key field type
# https://docs.djangoproject.com/en/4.1/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

LOGIN_REDIRECT_URL = '/'
EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'

### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###  
# API GENERATOR Specific
API_GENERATOR = {
    # API_SLUG -> Import_PATH 
    'product'  : "home.models.Product",
    'sales'    : "home.models.Sales",
}

REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': [
        'rest_framework.authentication.SessionAuthentication',
        'rest_framework.authentication.TokenAuthentication',
    ],
}
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### 

### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###  
# Dynamic DataTables Specific

DYNAMIC_DATATB = {
    # SLUG -> Import_PATH 
    'product'  : "home.models.Product",
    'sales'    : "home.models.Sales",
}

DYN_DB_PKG_ROOT = os.path.dirname( inspect.getfile( django_dyn_dt ) ) # <-- NEW App

STATICFILES_DIRS = (
    os.path.join(DYN_DB_PKG_ROOT, "templates/static"),
)
